### PR TITLE
feat(timepicker): add spinner support for TimePicker

### PR DESCRIPTION
**Pull request by enricoberti**
Needed for #854 

**Summary**

- The new DateTimePicker component requested on #854 needs some new components, like an inline calendar and a TimePicker with a numeric spinner. This PR takes care of the latter.

**Change List (commits, features, bugs, etc)**

- Overridden and extended TimePicker so we can increment hours and minutes with the spinner buttons and/or keyboard
- Works with RTL
- Doesn't throw any a11y error

image

**Acceptance Test (how to verify the PR)**

- new basic test suite available here: ?path=/story/watson-iot-experimental-timepicker--with-spinner

### CODE REVIEW

File: src/components/TimePicker/TimePicker.jsx [hunk 1]
```diff
@@ -0,0 +1,185 @@
+// export { TimePicker } from 'carbon-components-react';
```
[the remark this answers is not in the document]
Comment (enricoberti) [reply src/components/TimePicker/TimePicker.jsx]: thanks!

**Comment by enricoberti** [comment]
@sstone2423 good observation! i updated it to take care of the flip. I checked other spinners around and the consensus seem to be not to interact with incrementing hours when you perform a flip of the minutes, so I kept it simple and kept the flips independent.

Also, remove the useEffect to highlight the portion of time that is currently selected - that messed up a lot with manually writing the time

File: src/components/TimePicker/TimePicker.story.jsx [hunk 2]
```diff
@@ -1,3 +1,47 @@
...
+
 export {
   default as TimePickerStory,
 } from 'carbon-components-react/lib/components/TimePicker/TimePicker-story';
+
+storiesOf('Watson IoT Experimental/TimePicker', module)
+  .add('Default', () => {
+    return (
+      <div>
+        Please refer to the{' '}
```
Comment (tay1orjones) [inline src/components/TimePicker/TimePicker.story.jsx]: We need to copy the stories from carbon and paste them in here. The library could be using a different version of Carbon than what is present in their storybook, the component signature could differ from what we offer.

**Comment by tay1orjones** [comment]
@lukefirth this seems like a piece of functionality that is well isolated and would be a great candidate to push into Carbon. Did we discuss adding this to Carbon?

**Comment by tay1orjones** [comment]
After a discussion with @davidicus, I think an alternative approach that will allow us to avoid overriding the Carbon `TimePicker` would be ideal. `TimePicker` takes `children` as a prop, and places it at the same place in the markup that you have configured the "spinner" markup to render. 

The first change would be to isolate the spinner arrows to their own presentational `SpinnerControl` component that we could compose as a child of `TimePicker`.

From here we have two options:

1. Export a `TimePickerSpinner` component that pre-configures `TimePicker` `children` to be the `SpinnerControl` and would hold the stateful logic controlling the stock `TimePicker` via the click/key events from `SpinnerControl`. This component would probably render something simple like:
```jsx
<TimePicker>
  <SpinnerControl/>
<TimePicker>
```
2. Export a [custom hook](https://reactjs.org/docs/hooks-custom.html) to contain the stateful logic. The hook would accept the parameters required for the stateful event logic - the TimePicker ref, a ref for the SpinnerControl, events etc.
With this option we could provide documentation/story to demonstrate how to utilize this, but in sense it would be
```
import React from "react";
import { TimePicker, SpinnerControl, useTimePickerSpinner } from "carbon-addons-iot-react";

const App = () => {
  const { pickerValue } = useTimePickerSpinner({ timePickerRef ...whateverTheArgsNeedToBe });

  return (
    <TimePicker 
      ref={timePickerRef} 
      pickerValue
      ...etc
    >
      <SpinnerControl ...etc />
    </TimePicker>
  );
};
```

Either way, encapsulating this logic and using it as an augmentation to the Carbon component seems to be a much more sustainable approach than overriding the component. We can/should hop on a webex to chat, I realize this is quite the barrage of feedback for you today! 😅

**Comment by sstone2423** [comment]
> 3. The knobs are acting oddly - when I deselect the picker, knobs disappear, but when re-enabled the knobs do not reappear.
>    knobs-not-working

@tay1orjones  I've been noticing some odd behavior with other stories in that same fashion. There may be some leakage from other stories. Particularly the *Munich Building* title in stories that doesn't have that knob at all. May be a larger bug

**Comment by enricoberti** [comment]
> > 1. The knobs are acting oddly - when I deselect the picker, knobs disappear, but when re-enabled the knobs do not reappear.
> >    knobs-not-working
> 
> @tay1orjones I've been noticing some odd behavior with other stories in that same fashion. There may be some leakage from other stories. Particularly the _Munich Building_ title and the _isEditable_ knobs in stories that don't have those knobs at all. May be a larger bug. From the looks of it, they're knobs coming from `Card`

@tay1orjones @sstone2423 it seems actually they are a mix of knobs from [Dashboard](https://github.com/IBM/carbon-addons-iot-react/blob/master/src/components/Dashboard/Dashboard.story.jsx#L376) and [TileGallery](https://github.com/IBM/carbon-addons-iot-react/blob/master/src/components/TileGallery/TileGallery.story.jsx#L39) - Apparently other people [have it too](https://github.com/storybookjs/storybook/issues/1869)

**Comment by lukefirth** [comment]
Ok, responses to @tay1orjones 's comments. Take these in context as gut reactions of what I think perfect will be, and then we can back and forth.

----

> it is impossible to tell if I will increase the hour or minute value by pressing the up arrow. A reasonable remedy could be to place a slight underline below the currently selected time parameter (hours or minutes).
image

I like this approach. I would anticipate the interaction to be that these would only be visible when the spinner chevrons are being interacted with (hover, focus, click) and not otherwise (so don't show by default, don't show when typing inside the input etc).

That being said, this is a relatively low priority as immediately when a user interacts once, they get the feedback they need to understand what is happening, and the nature of this is that is wouldn't be slow or disruptive for them to recover from if it did do the opposite of what they expected.

-----

> The cursor should either stay put or not be visible when incrementing via up/down arrows. I'd like to ask for a visual/interaction review on these items.

Agreed, from what I can see in this PR and what looks like the implementation has been updated to do- not showing the cursor when interacting with the up/down chevrons looks to be the most natural way forward.


-----

I don't know where we ended with ease/ability to contribute back, but this would be a nice one if we can!

**Pull request merged**